src/index.js

```diff
@@ -1,9 +1,151 @@
 'use strict';
 
+const Ajv = require('ajv');
+const get = require('lodash.get');
+const set = require('lodash.set');
+
+const EXCLUDE_PATHS = new Set([
+	'serverless', 'serviceObject', 'pluginsData', 'package',
+	'configValidationMode', 'disabledDeprecations', 'serviceFilename', 'app',
+	'tenant', 'org', 'initialServerlessConfig'
+]);
+
+const SCHEMA = {
+	type: 'object',
+	properties: {
+		iterator: {
+			oneOf: [
+				{
+					type: ['array']
+				},
+				{
+					type: 'object',
+					oneOf: [
+						{
+							properties: {
+								$env: {
+									type: 'string'
+								}
+							},
+							required: ['$env'],
+							additionalProperties: false
+						},
+						{
+							not: {
+								required: ['$env']
+							},
+							additionalProperties: true,
+						}
+					]
+				}
+			]
+		},
+		template: {
+			type: ['array', 'object']
+		}
+	},
+	required: ['iterator', 'template'],
+	additionalProperties: false
+};
+
 class ForEachPlugin {
-	constructor(serverless, options) {
+	constructor(serverless) {
 		this.serverless = serverless;
-		this.options = options;
+
+		this.ajv = new Ajv({ allowUnionTypes: true });
+		this.validator = this.ajv.compile(SCHEMA);
+
+		this.hooks = {
+			'after:package:initialize': this.replace.bind(this),
+		};
+
+		if (this.serverless.configSchemaHandler && this.serverless.configSchemaHandler.defineFunctionEvent) {
+			this.serverless.configSchemaHandler.defineFunctionEvent('aws', '$forEach', SCHEMA);
+		}
+	}
+
+	interpolate(obj, key, value) {
+		const stringified = JSON.stringify(obj);
+		return JSON.parse(stringified
+			.replace('$forEach.key', key)
+			.replace('$forEach.value', value)
+		);
+	}
```

Comment:
```suggestion
	interpolate(template, key, value) {
		const stringified = JSON.stringify(template);
		return JSON.parse(stringified
			.replace('$forEach.key', key)
			.replace('$forEach.value', value)
		);
	}
```

At first glance I thought we were interpolating an object in the serverless object.

Reply:
Maybe add a try/catch and give a better error than a random json parsing error if the replace malforms the json in anyway.